### DISCUSSION TITLE
How can I send a notification/event from a JDTLS extension to the RedHat VS Code extension for broadcasting to another VS Code extension?

### DISCUSSION THREAD

**Discussion by guw** [redhat-developer/vscode-java#3520] [Q&A]
Here is my use case:

I have a JDTLS extension bundle jar, which gets plugged into JDTLS by the RedHat Java extension during JDTL initialization.
I now want to send an event from that extension via JDTLS back to VS Code. In VS Code I want to receive and process the event by some other extension, i.e. not within the RedHat Java extension.

Any ideas?

**Comment by rgrunber** [comment]
I had to look through this myself because I thought it was definitely possible, but we're just missing a small piece.

**From your JDT-LS extension bundle**
Use `JavaLanguageServerPlugin.getInstance().getClientConnection()`. It'll return you a `JavaClientConnection` which should contain all the method calls that go from server -> client. In particular the one you want is : [`sendEventNotification(EventNotification notification)`](https://github.com/eclipse-jdtls/eclipse.jdt.ls/blob/4b2b504bba98f6716dd90ac299b14015442dcedd/org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/JavaClientConnection.java#L167-L174)

That notification will travel from server to client.

**From your own VS Code extension**

vscode-java exposes its own api manager. See https://github.com/redhat-developer/vscode-java/blob/b041ca4f4db4ae921abc76e0bb1dbc0029cf19ee/src/apiManager.ts#L15-L81

I describe in https://github.com/redhat-developer/vscode-java/issues/3102#issuecomment-1561372221 how to access it, and there's quite a few extension that do it this way. Basically, the return value of activating the vscode-java extension is the API defined above.

```
const vscodeJava = extensions.getExtension('redhat.java');
const api = await vscodeJava.activate();
```

The problem is, we don't expose an emitter for any notification. Just some specific ones we care about, like errors. We could open this up. If you wanted a hack that might work immediately :

Instead of sending `sendEventNotification` like I described above, use [`telemetryEvent`](https://github.com/eclipse-jdtls/eclipse.jdt.ls/blob/4b2b504bba98f6716dd90ac299b14015442dcedd/org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/JavaClientConnection.java#L226-L232) instead. A telemetry event is like any other notification except it gets its own endpoint.

Our vscode-java extension will receive the telemetry event from the server and expose it to the API through : https://github.com/redhat-developer/vscode-java/blob/b041ca4f4db4ae921abc76e0bb1dbc0029cf19ee/src/standardLanguageClient.ts#L337-L338 .

Then just :

```
api.trackEvent((event: any) => {
...
...
});
```

I think this should work but haven't tried it myself. Otherwise, we should probably create an api event emitter for generic notifications.

**Reply by guw** [reply]
Would it be possible to expose the notifications in a similar way? We do have telemetry collection disabled in the preferences because we don't have approval for sending data to external systems.

Or alternatively, can we prevent `vscode-java` from sending events externally but sill have telemetry enabled?

**Reply by rgrunber** [reply]
I think so. I see it as the reasonable way forward. We should create a separate event transmitter for notifications from the server, and fire off the event to the api manager for all/most `languageclient.onNotification(..)` callbacks. @testforstephen , let me know if that seems reasonable, or if I've missed some existing way to get notifications through the api manager.

**Reply by testforstephen** [reply]
@guw You can use `JavaLanguageServerPlugin.getInstance().getClientConnection().sendNotification(clientCommandId, clientCommandParams)` to trigger a command on client side from server side. The clientCommandId is the vscode command you registered on your own extension, where you can either forward it to another extension or perform an action directly.

Here is the vscode-java side to handle such notification. https://github.com/redhat-developer/vscode-java/blob/7f89debd1508fb82511baf0f977ae1e27c0aa0c4/src/standardLanguageClient.ts#L313-L314
